tidy3d/components/simulation.py

```diff
@@ -2120,24 +2120,25 @@ def get_eps(structure: Structure, frequency: float, coords: Coords):
 
         def make_eps_data(coords: Coords):
```

Comment:
generalized this a bit over coords
```py
        def make_eps_data(coords: Coords):
            """returns epsilon data on grid of points defined by coords"""
            arrays = (np.array(coords.x), np.array(coords.y), np.array(coords.z))
            eps_background = get_eps(
                structure=self.background_structure, frequency=freq, coords=coords
            )
            shape = tuple(len(array) for array in arrays)
            eps_array = eps_background * np.ones(shape, dtype=complex)
            for structure in self.structures:
                # Indexing subset within the bounds of the structure
                # pylint:disable=protected-access
                inds = structure.geometry._inds_inside_bounds(*arrays)

                # Get permittivity on meshgrid over the reduced coordinates
                coords_reduced = tuple(arr[ind] for arr, ind in zip(arrays, inds))
                
                red_coords = Coords(**dict(zip('xyz', coords_reduced)))
                eps_structure = get_eps(structure=structure, frequency=freq, coords=red_coords)

                # Update permittivity array at selected indexes within the geometry
                is_inside = structure.geometry.inside_meshgrid(*coords_reduced)
                eps_array[inds] = eps_structure * is_inside

            coords = dict(zip('xyz', arrays))
            return xr.DataArray(eps_array, coords=coords, dims=("x", "y", "z"))
```

note, this and the code snippet from the `inside_meshgrid` that I wrote above seem to pass test for me locally.